### DISCUSSION TITLE
[iOS] 接触通知が異常な割合でバッテリーを消費するとの報告がある

### DISCUSSION THREAD

**Discussion by keiji** [cocoa-mhlw/cocoa#248] [Ideas]
## 不具合の内容 / Describe the bug

Twitterで「接触通知が異常な割合でバッテリーを消費することがある」ことが複数件、報告されているのが気になっています。

> 過去24時間のバッテリー消費の半分が「接触通知」ってビックリだ。iOSは14.6、機種はiPhone8。ちなみに #COCOA アプリは1.2.4

（https://twitter.com/shadomobile/status/1404799904002805774 より）

ツイートに添付されている画像には「接触通知」が52%のバッテリーを消費しているとして、バッテリー消費ランキングの1位に表示されています。

「接触通知」なので本来はCOCOAではないのだけど、COCOAの不具合のように語られてしまう（COCOAを使ってるから接触通知が働くという意味では間違いではない

## 再現手順 / Steps to reproduce

再現手順が不明。見つけ出せればAppleに問い合わせするハードルが下がりそう。

## 期待される挙動 / Expected behavior

BLEだし、そこまでバッテリーを消費するものでもないと思うので、少なくともバッテリー消費ランキングトップに出てこないことを期待している。

## スクリーンショット / Screenshots

前述のツイートを

## 動作環境 / Environments

- デバイス：iPhone 8 で報告、他にも発生する可能性がある
- OS： iOS 14.6 で報告、他にも発生する可能性がある
- バージョン：1.2.4 で報告、他にも発生する可能性がある

## その他 / Additional context

手がかりに乏しいので、とにかく事例を集めて、共通項がないか検討するのが良さそう。

**Comment by yoshitomo-g** [comment]
同様のツイートはかなり早い段階から（おそらくv1.2.0以前）頻繁に見かけるのですが、私の端末では何の問題もないのでずっと気になっています。

情報提供ということで、放置端末の充電率の下がり具合です。この時の接触確認アプリはv1.2.1でしたが、v1.2.4になった現在も大きな違いはなさそうです。
https://twitter.com/yoshitomo_y/status/1401106510667599877

消費ランキングですが、全消費の中での割合でしかないので、上位だとしてもどれくらい消費しているのかまではわかりません。該当期間でどれくらい充電を消費したかの情報もセットで提供してもらう必要があると思います。
ワシントン州版アプリでは同様の説明をしていて、そのことについてもツイートしています。
https://twitter.com/yoshitomo_y/status/1379821811106385924

---
以下は私の考察です。検証はしていません。

私はバッテリーの情報を頻繁に見るのですが、ランキングに「接触確認アプリ」が表示されることはほとんどありません。それに対して「接触通知」は高頻度かつ上位に表示されます。バックグラウンドの処理が表示される確証はないのですが、フォアグラウンドとバックグラウンドを区別した表示になっているので、バックグラウンドの場合でも表示されるのだろうと思っています。これを前提にするなら、実際に消費しているのは「接触通知」のほうと考えて良さそうです。

内訳としてはこんなところでしょうか。

- RPI配信
  - RPI作成・保存
  - BLE送信
- RPI受信
  - RPI保存
  - BLE受信
- 接触チェック
  - RPI読み込み
  - 診断キーからのRPI作成
  - RPI比較
  - 接触チェックログの更新

この中では、RPI受信と接触チェックの２つが端末によって違いが出そうです。アプリを入れている端末が周囲に多ければ受信も多く、RPI比較の回数も増えるのではと考えました。この点では、前述の放置端末は参考にはならないかもしれません。

充電消費を1時間ごとに確認すると傾向が出て、次のような判断ができるのではと想像しています。手間がかかるので、お願いするのは難しいと思いますが・・・。

- 人の多い場所にいた時間だけ消費が増える → RPI受信
- 深夜などに1回だけ大きく消費する → 接触チェック

**Comment by i-maruyama** [comment]
同じ問題について Android でも yoshitomo-g さんと同様にバッテリー消費見ていたのですが、有意義な情報を得るのが難しかったです。

提案しようとして時間がなかったのですが「接触通知をオフにしてアプリを閉じる」ボタン（外出時にはアプリを起動してくださいという注意書き付き）がどこかにあれば、ユーザーには親切だと思います。

**Comment by kvaluation** [comment]
バッテリー消費、他のアプリとの相対的な差で評価することで、原因を特定しやすいと思われます。

iPhoneでなくAndroidの場合で恐縮ですが、比較したいアプリとして下記があります。

■BeaconScope: ENを含めBLTを受信し電波強度から距離を測定している。
→BeaconScopeと連動して接触確認の消費電力が多いなら、BLT通信量（人混み）に比例してバッテリーを消費している可能性。
（iPhone版は発見できず）

■SilentLog: GPS信号を受信している。
SilentLogの方がCOCOAよりバッテリーを消費していたら、架空のGPS型の接触確認アプリと比較して、現在のBLTはバッテリー消費の面では利点がある。

きちんと実験はしていませんが、
SilentLogとGoogle Play開発者サービスが同程度、BeaconScopeはそれより消費電力が少ない感じなので、私としては、BLT通信の電力消費だけで決まっていない可能性もありえるかと思います。

すれ違ったCOCOAオンの端末数や、すれ違いの総経過時間に比例して、RPIの保存数や、近接にあった相手方の電波強度の平均値の算出処理の量が増加するので、実験環境を整えるのは難しそうです（自宅置きっぱなしと、繁華街おでかけの比較など）。

**Comment by b-wind** [comment]
例示されているTweet、「探す」アプリ（Android での 「端末を探す」アプリと同様の機能）が 30% も占めてるし、トータルのバッテリー消費が少なすぎて相対的に上になってしまっただけのような……？

直接関係ないけれど、参考情報。
https://japanese.engadget.com/ios146-battery-drain-test-020054341.html

（バッテリー消費が激しい問題、毎年のように繰り返してくれるので気にしなくなってしまった）

**Comment by b-wind** [comment]
そもそも「接触通知」のバッテリー消費の内訳が謎なのでその辺りは注意して考える必要はあると思います。
（例えばBluetooth で消費した電力が含まれるのか含まれないのか明示されてないはず）

**Comment by Takym** [comment]
「接触通知」と書かれており「COCOA」とは書かれていませんので、アプリではなくOS側の原因である気もします。

**Comment by Takym** [comment]
@yoshitomo-g さん

> 私はバッテリーの情報を頻繁に見るのですが、ランキングに「接触確認アプリ」が表示されることはほとんどありません。それに対して「接触通知」は高頻度かつ上位に表示されます。バックグラウンドの処理が表示される確証はないのですが、フォアグラウンドとバックグラウンドを区別した表示になっているので、バックグラウンドの場合でも表示されるのだろうと思っています。これを前提にするなら、実際に消費しているのは「接触通知」のほうと考えて良さそうです。

「接触通知」とはCOCOAのバックグラウンド処理が分けて表示されているだけという事でしょうか。僕はOS側のプログラムの事だと思いました。

**Comment by b-wind** [comment]
> 「接触通知」なので本来はCOCOAではないのだけど、COCOAの不具合のように語られてしまう（COCOAを使ってるから接触通知が働くという意味では間違いではない

中身を明確に定義したドキュメントがあるかどうか分かりませんが、所謂 EN API の本体側なのだと推測されますね。

**Comment by Takym** [comment]
- 「接触通知」は COCOA の一部では無い事
- 人が多い場所では電気を多く消費する可能性が有り得る事
- 電気の使用量は相対的な数値で有る事

この３点を認知して貰う事が重要ですね。

**Comment by b-wind** [comment]
後は BLE も接触通知の為だけに動いて居るわけでは無いと言う点とかですね。
AirDrop なんかも既定で有効の筈ですし、他にも様々な用途で使われています。

**Comment by keiji** [comment]
ありがとうございます。最終的にAppleに尋ねることになると思いますが、その前に事例の収集と共通項の洗い出しをして、可能であれば再現条件を探りたいというのがこのIssueの目的です。

相対的な利用割合に過ぎないという考え方はある一方で、グラフの傾斜として確かにバッテリーの減り方が著しいなと思えるケースも見たことがあります。

また「バッテリーを消費しているのは『接触通知』であってCOCOAではない」というのも技術的には同意します。

その上で、忘れてはいけないのは「利用者からすればそんなことは関係ない」と言うことです。COCOAを使っていて（接触通知が有効になっていて）バッテリーが減るという事実に変わりはなく、スマートフォンユーザーにとってバッテリーを消費するアプリは悪者なので、COCOAがアンインストールされる一因になることは可能な限り防ぎたい。ぼくはそう考えています。

周りにたくさんの端末があり、RPIの受信と保存処理に時間がかかっているという点に関しては面白い観点だと思います。原理的な要因であればAndroid端末でも同様の報告があると思うのですが、バッテリー消費に関する報告は基本的にiOSに集中しているように思うのが気にかかっています。これについてはある端末の周辺にENを発信する端末が1台のケースと、５台のケースで、それぞれバッテリーの消費に違いが起こるか計測するなどで検証できそうですね。

手が空き次第、Twitterなどで類似の報告があるか確認して本Issueに集約していく予定です。

**Comment by b-wind** [comment]
> すれ違ったCOCOAオンの端末数や、すれ違いの総経過時間に比例して、RPIの保存数や、近接にあった相手方の電波強度の平均値の算出処理の量が増加する

検証可能という点では確かにこれなんですよね。
端末固有のバグか、iOS 自体のバグかなど、さらに条件がある可能性もありますが。

**Comment by b-wind** [comment]
余談ですが、初期の実装だと「COVID-19接触のログ記録」という形で「ヘルスケア」アプリのバッテリー消費が大きく表示される事はあったようで。

**Comment by b-wind** [comment]
実はCOCOAが原因だった場合はここぐらい。（いろいろ考えて指定されていると思いますが、実際の意味合いは私にはよくわからない）
https://covid19radar-jpn-prod.azureedge.net/c19r/Configration.json

~~https://developer.apple.com/documentation/exposurenotification/enexposureconfiguration~~
https://developer.apple.com/documentation/exposurenotification/enexposureconfiguration/exposure_risk_value_calculation_in_exposurenotification_version_1
（単に関係ありそうな所を列挙しているだけでどれも根拠はないです）

**Comment by yoshitomo-g** [comment]
@kvaluation 
> ■BeaconScope: ENを含めBLTを受信し電波強度から距離を測定している。
> →BeaconScopeと連動して接触確認の消費電力が多いなら、BLT通信量（人混み）に比例してバッテリーを消費している可能性。
> （iPhone版は発見できず）

iOSだと、BLE Scanner が近いと思います。iOSの制限で利用者の調査には使えませんが、BLEで発信している端末数の確認はできるようです。
https://apps.apple.com/jp/app/id1221763603

**Comment by yoshitomo-g** [comment]
@Takym 
> 「接触通知」とはCOCOAのバックグラウンド処理が分けて表示されているだけという事でしょうか。僕はOS側のプログラムの事だと思いました。

バックグラウンド云々は、「接触確認アプリ」がバックグラウンドで動いていても表示されるだろうから、上位に表示されないのならバッテリーの消費はわずかだろうという推測の話です。

「接触通知」がOS側という認識については、私もその通りだと思います。

**Comment by b-wind** [comment]
手元の iPhone だとこんな感じですね。

過去24時間

- 「接触通知」： 2%
- 「設定アプリ」： 1%  （一例として）
- 「接触確認アプリ」：「ー」表記（1% より下）

**Comment by keiji** [comment]
https://twitter.com/cheepao/status/1405133228831170570

> iPhone SE2 をiOS 14.4.2で利用していますが、いつも6割近くのバッテリーを消費しています。
昨年10月までiPhone 8を使ってましたが、同様でした。
関係あるかは不明ですが、Bluetooth でFitbitを接続して利用しています。
以上ご参考になれば幸いです。

**Comment by keiji** [comment]
https://twitter.com/pmakino/status/1405135470560878592

> 時たま発生してるんだけど (と言っても4時間で20%食うとかその程度だけど)、いつどんなときに起きるのかよくわからなくてレポートしがたいんだよね…

4時間で20%... ~~１時間あたり5%（と言う考え方でいいのかな？）~~ が正常の範囲内かはわかりませんが、引き続き情報を集めます。

----
https://twitter.com/pmakino/status/1405141804618436608

> iPhone SE 第二世代、iOS 14.6 です。
なお直近24時間の電力消費内訳はこんな感じでした。今日の総バッテリー消費量はざっくり30%程度で、接触通知はそのうちの7%なので大したものではないですが、比較的よく使った(数十分程度?)Janetterより多いのは若干違和感あります。

----
https://twitter.com/pmakino/status/1405142595152465924

> また直近10日の場合はこちらで、これもLINE等より接触通知が多いのは違和感あります。とはいえこれらは異常と感じる日のものではなく、「なんじゃこりゃ」と異常に気付く日は内訳の断然トップに接触通知が来ます。(70%とか)

**Comment by keiji** [comment]
https://twitter.com/nut320/status/1405131346352087044

>トップじゃないけど2位で、確かに異様にバッテリー消費多い気がする。アプリの画面なんてさっき数ヶ月ぶりに10秒くらい開いただけなのに。
iOS 14.6、cocoa 1.2.4
> 
> ツイッターも食い過ぎな気がするけど……

**Comment by keiji** [comment]
https://twitter.com/SPRIT_BRAIN/status/1405124852680323081

> iphone7 ios14.6で運用していると、何もしなくても2日目の夕方にはバッテリーが20%近くになってる。
うちの２台ともそうだから、これかなぁ。気になる。

ある程度情報を集めたら、「バージョン」「機種」のグリッドにしたい

**Comment by Takym** [comment]
僕の環境では接触通知は19%でしたが、他のアプリ（４つほど）も10〜20%台でした。バージョンは iOS 14.6 です。COCOA本体は表示されていませんでした。

## 追記
上記の値は24時間以内の使用率です。10日分の接触通知の使用率は10%で、動画配信アプリが31%、ブラウザが17%を占めていました。

**Comment by Takym** [comment]
~~そもそも「バッテリー使用状況」の値は単位からは「使用総量」なのか「合計使用量に対する割合」なのかよく分かりませんね。画面右上の電池残量は「％」で表示されていますし、「バッテリー使用状況」も「％」で表示されています。~~

設定アプリのバッテリページの下部に「使用率」と明記されていました。

**Comment by Takym** [comment]
@keiji さん

> 4時間で20%...１時間あたり5%（と言う考え方でいいのかな？）が正常の範囲内かはわかりませんが、引き続き情報を集めます。

4時間の間に消費した量が20%という事ですから、「1時間あたり5%」では無いと思います。

**Comment by yoshitomo-g** [comment]
Appleの説明（[iPhone、iPad、iPod touch のバッテリーの使用状況について](https://support.apple.com/ja-jp/HT201264)）には次のようにあります。
> 画面上の棒グラフでいずれかの棒をタップすると、その時間帯にバッテリーをいちばん消費していた App や、その App のバッテリー消費量の全体に占める割合を確認できます。

---
常用している端末の24時間の内訳は、現時点で次のような感じです。使用時間は、ニュースアプリが90分、ブラウザアプリが1時間弱くらいです。

- ニュースアプリ　28%
- ブラウザアプリ　28%
- 接触通知　15%
- Twitter　13%
- （中略）
- 設定　1%
- 接触確認アプリ　1%以下

**Comment by b-wind** [comment]
1時間単位で見てみると、接触通知 100% の時間帯は見られますね。
バッテリー残量が十分にあって、（多分）寝てる時間だったかと。

**Comment by b-wind** [comment]
COCOA インストール後のバッテリー消費（実測）に触れている記事。（どれも去年のデータ）
個人的印象としては思ったよりは少々大きめですかね……

https://www.mdn.co.jp/di/contents/4854/74163/
https://carbon-freelance.net/2020/06/cocoa-apps-battery-consumption-rate.html
https://moduleapps.com/mobile-marketing/18413app/

**Comment by b-wind** [comment]
なるほどー。
一枚目のスクショを取った直後に電池不足で落ちて充電しながらスクショ取ったのが2枚目なので、電池の残量とかあてにならないなと思いました。まる。

<div>
<img src="https://user-images.githubusercontent.com/144311/122312280-d998ca80-cf4e-11eb-9a78-664f41997b5f.jpg" width="30%" >

<img src="https://user-images.githubusercontent.com/144311/122312863-0bf6f780-cf50-11eb-8718-f51a72eac7ca.jpg" width="30%" >

</div>

**Comment by b-wind** [comment]
あくまで手元端末の観測結果ですが、充電中だと「接触通知」は見られず充電器から外して放置してある状態で「接触通知」のバッテリー消費割合が高く表示されます。
前述のようにバッテリー残量の値が信用なら無いので「接触通知」の電力消費量自体は不明です。

**Comment by tatsu-jp** [comment]
過去10日間の棒グラフを、以下の見方で分析しています。
- 一日ごとのバッテリーの使用状況：実際に使用した量（100%＝バッテリーをフルに使い切った）
- Appのバッテリー使用状況の内訳：棒グラフの使用分の内訳

＃仕様を見つけられていないので、見方が合っているかは確証持てていません。。。
＃日中に充電すると100%の意味合いが変わってきそうですが、私は日中充電していないのでとりあえず無視してます。

上で合っているとして話を進めると、例えば、棒グラフ上でバッテリー使用状況が30%、内訳で接触通知が50%使用していた場合、

接触通知が一日で使用したバッテリー量 = 30% * 0.5 = 15%

とざっくりと計算できます(棒グラフでは概算値しか取れないのですが、大まかな値が取れれば十分かと思っています)。
これで計算できるのであれば、私の端末(iphone11/ios14.6)では過去10日間で、接触通知がおよそ2〜8%(/日)のバッテリーを消費しています。ステイホーム時は2~3%、外出時は5%以上になっていることから通信量に応じて電力消費しているのかな？と漠然と考えています(推測の域ですが)。
私はこんなものなのかなと思って気にしていませんでしたが・・・バッテリー容量の小さい端末の場合は同じ処理でも残量(パーセンテージ)の減る速度は早いので、消費量が大きいと捉えられるケースが多いのかもしれないと思いました。

**Comment by yoshitomo-g** [comment]
https://twitter.com/bishibashi/status/1407136452270465027
> 一晩でバッテリ切れになったのは接触確認アプリのせい、ってこと？
> 急にバッテリ喰うようになったような。

詳しく話を聞いてみました。
https://twitter.com/yoshitomo_y/status/1407267616712916994

端末：iPhone X
OS：iOS 14.6（アップデートから30日ほど経過）
アプリバージョン：1.2.4
バッテリー状態：劣化有り（最大容量89%）

**Reply by kvaluation** [reply]
> 詳しく話を聞いてみました。

よしともさんありがとうございます。

iOS 14.6へのアップデートとの相関関係はなさそうと理解しました。
COCOA 1.2.4 のアップデートで、iOSの接触確認に何か無駄に負荷のある処理が発生した可能性とかも想定してみた方が良いのか、謎が深まりました。

手元の使っていないiPhoneで、GPSをとるSilentLogの常時起動より接触確認のバッテリー使用率が大きいので、BLEではない接触確認内部の負荷かもという、keijiさんの仮説もありえると、思います。

**Reply by kvaluation** [reply]
## GPSを利用する常駐アプリとの比較

### 図１　アクティビティー

<img src="https://user-images.githubusercontent.com/60703087/123288973-fcc50a80-d54a-11eb-8653-68a2580d7e68.PNG"  width="350">



### 図２　バッテリー使用状況

<img src="https://user-images.githubusercontent.com/60703087/123289413-6c3afa00-d54b-11eb-979f-1c564ebc4db2.PNG"  width="350">


### 使用状況
1. iPhone 9.1, iOS 14.3, COCOA 1.2.4, build_number 1622417525
2. iPhoneはずっとおうちでお留守番。外出せず、他のiPhone / Android数台と近接。
3. ログ当日のcocoa-log
https://gist.github.com/kvaluation/b03486701e1af89c2b9c7b55c837e130#gistcomment-3791736

### SilentLogとの比較
SilentLogは、GPSを使う常駐型のライフログアプリで、BLEよりも電力を消費する可能性がある。
しかし、圧倒的に接触確認の消費電力が大きかった。

### BLE Scanner
@yoshitomo-g さん
> iOSだと、BLE Scanner が近いと思います。iOSの制限で利用者の調査には使えませんが、BLEで発信している端末数の確認はできるようです。
> https://apps.apple.com/jp/app/id1221763603

ありがとうございます。短時間だが、BLE Scannerを立ち上げて、近接の端末とのBLE信号強度を確認。
しかし、BLE Scannerの電力消費はほぼなし。BLE信号の受信で消費電力が大きくなるということはなさそうであることが示唆された気がする。

### 相対的な値の価値
バッテリーの残量測定は、電圧と温度と事前測定値によるなど、あまり精度が高くないと思われますが、同一機種内での相対的な評価については、一定程度の信頼性があるかと思われます。
なにか、接触確認をこのアプリと比べてみよう、という実行例がありましたら教えてください。

### おでかけとの比較
iPhoneのお留守番とおでかけで比較できると良いのですが、着手できてません。

**Reply by keiji** [reply]
@kvaluation
ありがとうございます。
yoshitomo-gさんのコメントにリプライするかたちになっているので、次からDiscussionに直接コメントしてもらえるとうれしいです。

**Comment by keiji** [comment]
来週からになりますが、個人的に次の実験をする予定です。

 * iOS端末を満充電した後、バッテリー駆動に切り換えてn時間放置したあと「接触通知」が残バッテリー容量（消費しているバッテリー量）を確認する

これを、端末の周囲に置く接触確認APIを有効にした端末（Android端末）の数を「0個, 1個, 3個, 5個」に変えてそれぞれ計測します。

時間をnにしているのは、今ぼくが持っているiOS端末がかなり古いiPhone SE2で、バッテリーがへたっているので、充電していない状態では1時間持たない可能性があるからです。比較的新しいiOS端末を投入できるか調整してみます。

**Reply by keiji** [reply]
実験結果です。

まずはiPhoneの周りにEN端末が0台置いたケースと、5台置いたケースで、それぞれ6時間後の消費バッテリー量を比較しましたが、どちらも7%の消費と、大きな差はありませんでした。

端末：iPhone 6s
OS：iOS 14.6
アプリバージョン：1.2.4


## 0台

![IMG_0004](https://user-images.githubusercontent.com/932136/124406402-3b916680-dd7c-11eb-9e6a-5c1160f8b17d.PNG)

## 5台

![PXL_20210704_104334295](https://user-images.githubusercontent.com/932136/124406380-2e747780-dd7c-11eb-92ea-62ecd005785e.jpg)
![IMG_0007](https://user-images.githubusercontent.com/932136/124406393-346a5880-dd7c-11eb-95e9-0769e9c95031.PNG)

**Comment by keiji** [comment]
端末：iPhone12 Pro Max
OS：iOS 14.6
アプリバージョン：1.2.4

過去10日間で3日前までは正常。突然バッテリー消費が激しくなる。屋内で利用。
「おやすみモード」の利用をしたことがあるとの付記事項あり。

https://twitter.com/yoko_mind/status/1410944545961910272

**Comment by keiji** [comment]
iOS端末が受信側での実験（ https://github.com/cocoa-mhlw/cocoa/discussions/248#discussioncomment-963905 ）は今のところ空振りなので、アプローチを変えてiOS端末を発信側としてバッテリー消費とBLEのアドバータイジングの関係について見てみます。

![PXL_20210705_133929884](https://user-images.githubusercontent.com/932136/124534548-71534f80-de4f-11eb-8c47-a098f0f35edf.jpg)

iOS端末が発信するアドバータイジングをBluetoothスニッファ（代わりのRaspberry PI）で記録します。バッテリーの異常な消費が観測できれば、その期間のアドバータイジングの記録と照らし合わせます。

BLEのアドバータイジングの発信が異常な頻度で行われているなどの現象が観測できれば、原因が絞り込めるのではないかと考えています。

ちなみに写真ではシールド袋に入れていますが、実際の実験は、他の端末がない離れた部屋にiOS端末とRaspberry Piを放置して行っています（シールド袋に入れてもWi-Fiでガンガン通信していたのだけど、これは5GHz帯だからかな…）。
